daemon/k8s_watcher.go

```diff
@@ -913,22 +920,33 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
 				},
 				k8s.ConvertToNode,
 			)
-			isConnected := make(chan struct{})
-			// once isConnected is closed, it will stop waiting on caches to be
+			stopController := make(chan struct{})
+			// once stopController is closed, it will stop waiting on caches to be
 			// synchronized.
-			d.blockWaitGroupToSyncResources(isConnected, nodeController, k8sAPIGroupNodeV1Core)
+			d.blockWaitGroupToSyncResources(stopController, nodeController, k8sAPIGroupNodeV1Core)
 
 			once.Do(func() {
 				// Signalize that we have put node controller in the wait group
 				// to sync resources.
 				asyncControllers.Done()
 			})
 			d.k8sAPIGroups.addAPI(k8sAPIGroupNodeV1Core)
-			go nodeController.Run(isConnected)
-			// TODO: do we really need to wait until etcd watcher signalizes
-			// "listDone" or connected to it is sufficient?
-			<-kvstore.Client().Connected()
-			close(isConnected)
+			go nodeController.Run(stopController)
+
+			// Wait for kvstore to be connected and consider
+			// switching to kvstore based event handling if the
+			// optimization is enabled
+			for {
```

Comment:
Why have an infinite for loop here? Couldn't we wait until the client is connected, then have something like 

```
<-kvstore.Client().Connected()
if !option.Config.K8sEventHandover {
    return
}
```


Then have the same logic as before?

Reply:
Great question. I've added the loop so we can change the option at runtime. I excluded this in the scope of 1.5 but I want to add logic in 1.6 to enable this option (unless disabled by the user) as we pass 200 nodes in the cluster.

Reply:
I don't fully understand this, we're already in an infinite loop so we would already do this retrying. In that scenario, though, we would also bootstrap again via k8s. If we wanted to avoid that I'm not sure how this code does that, since it breaks the inner infinite loop, but would then fall down to `<-kvstore.Client().Disconnected()` and then loops again.
Since this is a scale issue, would allowing this watcher to connect ever make sense?

Reply:
@tgraf by "change the option at runtime" do you mean changing the configuration dynamically without restarting the cilium-agent? Is this possible via cobra / viper? Or rather that we would change the configuration dynamically if we scale up to over 200 nodes?

What about the scale-down case? I.e., what if we change from `option.Config.K8sEventHandover = false` to `option.Config>k8sEventHandover=true` - we will never enter back into this logic if we then set it back to `option.Config.K8sEventHandover = false` since we break out of the loop here. This means that once we rely on the kvstore, we always rely on it - which isn't a bad thing, just confirming that this is the desired behavior.

Reply:
> by "change the option at runtime" do you mean changing the configuration dynamically without restarting the cilium-agent? Is this possible via cobra / viper? Or rather that we would change the configuration dynamically if we scale up to over 200 nodes?

Yes, this. Given that this seems non-obvious, I'll change this PR to not include this and convert it into a loop in the context of 1.6. It will be easier to review with all context.





> What about the scale-down case? I.e., what if we change from `option.Config.K8sEventHandover = false` to `option.Config>k8sEventHandover=true` - we will never enter back into this logic if we then set it back to `option.Config.K8sEventHandover = false` since we break out of the loop here. This means that once we rely on the kvstore, we always rely on it - which isn't a bad thing, just confirming that this is the desired behavior.

So far, the plan has been to never revert back unless cilium is restarted after scale-down has occured. Let's discuss when I propose the PR specifically. I think both options are correct.

Reply:
Changed.